src/space_usage.rs

```diff
@@ -0,0 +1,110 @@
+    pub largest_webxdc_data: Vec<(usize, usize, usize)>,
+}
+
+impl std::fmt::Display for SpaceUsage {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        writeln!(f, "SpaceUsage:")?;
+        let human_db_size = format_size(self.db_size, BINARY);
+        writeln!(f, "[Database Size]: {human_db_size}")?;
+        writeln!(f, "[Largest Tables]:")?;
+        for (name, size, row_count) in &self.largest_tables {
+            let human_table_size = format_size(*size, BINARY);
+            writeln!(
+                f,
+                "   {name:<20} {human_table_size:>10}, {row_count:>6} rows",
+                name = format!("{name}:"),
+                row_count = row_count.map(|c| c.to_string()).unwrap_or("?".to_owned())
+            )?;
+        }
+        writeln!(f, "[Webxdc With Biggest Status Update Space Usage]:")?;
+        for (msg_id, size, update_count) in &self.largest_webxdc_data {
+            let human_size = format_size(*size, BINARY);
+            writeln!(
+                f,
+                "   {msg_id:<8} {human_size:>10} across {update_count:>5} updates",
+                msg_id = format!("{msg_id}:")
+            )?;
+        }
+        Ok(())
+    }
+}
+
+/// Get space usage information for the Context's database
+pub async fn get_space_usage(ctx: &Context) -> Result<SpaceUsage> {
+    let page_size: usize = ctx
+        .sql
+        .query_get_value("PRAGMA page_size", ())
+        .await?
+        .unwrap_or_default();
+    let page_count: usize = ctx
+        .sql
+        .query_get_value("PRAGMA page_count", ())
+        .await?
+        .unwrap_or_default();
+
+    let mut largest_tables = Vec::new();
+
+    let biggest_tables = ctx
+        .sql
+        .query_map_vec(
+            "SELECT name,
+                SUM(pgsize) AS size
+                FROM dbstat
+                WHERE name IN (SELECT name FROM sqlite_master WHERE type='table')
+                GROUP BY name ORDER BY size DESC LIMIT 10",
+            (),
+            |row| {
+                let name: String = row.get(0)?;
+                let size: usize = row.get(1)?;
+                Ok((name, size))
+            },
+        )
+        .await?;
+
+    for (name, size) in biggest_tables {
+        let row_count: Result<Option<usize>> = ctx
+            .sql
+            // SECURITY: the table name comes from the db, not from the user
+            .query_get_value(&format!("SELECT COUNT(*) FROM {name}"), ())
+            .await;
+        largest_tables.push((name, size, row_count.unwrap_or_default()));
+    }
+
+    let largest_webxdc_data = ctx
+        .sql
+        .query_map_vec(
+            "SELECT msg_id, SUM(length(update_item)) as size, COUNT(*) as update_count
+                 FROM msgs_status_updates
+                 GROUP BY msg_id ORDER BY size DESC LIMIT 10",
+            (),
+            |row| {
+                let msg_id: usize = row.get(0)?;
```

Comment:
You can convert this directly into `MsgId`:
```suggestion
                let msg_id: MsgId = row.get(0)?;
```